### DISCUSSION TITLE
Setting up OIDC and TLS using the java agent?

### DISCUSSION THREAD

**Discussion by DGuhr** [open-telemetry/opentelemetry-java-instrumentation#10548] [Q&A] [answered]
Hi, sorry to bug but I have another question: Is it possible to use oidc / oauth2 client config and tls when using the java agent to send spans with transport encryption?

I have read this article: https://medium.com/opentelemetry/securing-your-opentelemetry-collector-1a4f9fa5bd6f 

and wanted to try it out in a docker setup with a slightly different environment. Currently at the stage where everything seems to work, but my apps are not able to send spans or metrics (which is shown in the logs). 

I have the following setup on the client, which is a spring boot 2 app as a fatjar

Multi-Stage Dockerfile: 
```
# Step 1: OpenTelemetry Java Agent
FROM curlimages/curl:latest AS downloader
WORKDIR /app
RUN curl -L -o opentelemetry-javaagent.jar "https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/download/v1.32.1/opentelemetry-javaagent.jar"

# Step 2: Build the application
FROM openjdk:17-jdk-slim

WORKDIR /app

COPY --from=downloader /app/opentelemetry-javaagent.jar opentelemetry-javaagent.jar
ADD target/*.jar app.jar
COPY agent-exporter /etc/otelagent

EXPOSE 8080

CMD ["java", "-javaagent:opentelemetry-javaagent.jar","-Dotel.javaagent.configuration-file=/etc/otelagent/otelexporterconfig.properties", "-jar", "app.jar"]
```
 the copied directory `agent-exporter` contains the files `ca.pem`, `ca-key.pem` and `client.pem` for SSL.

this is the configuration file content, which I expect to be wrongly formatted and would love to get help with: 
```
otel.extensions.oauth2client/client.client_id=agent
otel.extensions.oauth2client/client.client_secret=<secret-value>
otel.extensions.oauth2client/client.token_url=http://0.0.0.0:8080/realms/opentelemetry/protocol/openid-connect/token
otel.receivers.otlp.protocols.grpc.endpoint=localhost:5317
otel.exporters.debug=true
otel.exporters.otlp/secure.endpoint=otel-collector:4317
otel.exporters.otlp/secure.tls.ca_file=ca.pem
otel.exporters.otlp/secure.tls.cert_file=client.pem
otel.exporters.otlp/secure.tls.key_file=client-key.pem
otel.exporters.otlp/secure.auth.authenticator=oauth2client/client
otel.service.extensions=oauth2client/client
otel.service.pipelines.traces.receivers=otlp
otel.service.pipelines.traces.exporters=otlp/secure,logging
```

The whole service is actually part of a bigger docker-compose, where otel-collector is the external collector, and where e.g. service name and metrics config is set using environment variables. 

current log output is: 
```
2024-02-14T15:29:19.771202129Z 2024-02-14T15:29:19.770Z  WARN 1 --- [alhost:4317/...] i.o.exporter.internal.grpc.GrpcExporter  : Failed to export metrics. Server responded with gRPC status code 2. Error message: Failed to connect to localhost/127.0.0.1:4317
...
2024-02-14T15:29:24.800589923Z 2024-02-14T15:29:24.799Z  WARN 1 --- [alhost:4317/...] i.o.exporter.internal.grpc.GrpcExporter  : Failed to export spans. Server responded with gRPC status code 2. Error message: Failed to connect to localhost/127.0.0.1:4317

```

I think I either have something fundamentally wrong or I messed up the properties keys. Might be something else entirely, for sure. Anyways, help is highly appreaciated :) Thanks!

**Comment by laurit** [comment, marked as answer]
Collector and agent are configured differently, see https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md

**Reply by DGuhr** [reply]
thanks, this helps with TLS i guess.  But is it somehow possible to get a token using client id/secret using the agent? I can't see anything. I don't think so, right? Maybe MTLS?  Sorry, I really try to understand it, that'S all there is. that linked articles usage of the word "agent" (e.g. in "config.agent.yaml") confused me. What I did so-far was just changing that yaml to properties, but I already guessed that was not it.

**Reply by laurit** [reply]
Otel java agent offers an extension mechanism https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension that lets you customize the agent. Firstly see https://github.com/open-telemetry/opentelemetry-java/pull/4630 This should let you add headers to the requests, idk whether this is enough to implement oauth2.

**Reply by DGuhr** [reply]
Thanks, i'll look at this again tomorrow. I think(!) I had an understanding mistake. The article talks about 2 collectors if I read it now right, one is the oauth client, using this collector extension: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/oauth2clientauthextension/README.md it reaches out to get a token via clientid/clientsecret and sends the token when spans are exported, and the other collector is the relying party, which checks the token validity using the configured audience claim aud. 

I thought, reading the article, that there was a way to enable the first part - talking to an IdP directly from the java agent to obtain a token and only with this token as authorization header present the spans are sent, and further only when the valid aud claim is evaluated in the collector they are exported by the collector to a metrics / tracing /... backend - was possible. It seems not, without writing custom extensions.

This would allow to kinda "whitelist" span exporting per service using oauth client credentials flow (RFC 6749) and validatiom, which might be a valid usecase. I still do not quite get the usecase from the article, then, but that may be just me. It talks about an authentication proxy, which is fine, but when this collector lives in a public network there's still attack surface when I somehow find out its dns and push signals to it, right?  Imo it would make more sense to create access control at the point where signals are actually emitted, which is the app/agent/... - but I may just understand the article wrong. Nevertheless thank you for making me aware of the different config format for agent/exporter, so given the above is right, I think we can leave it as is here for now.

**Reply by DGuhr** [reply]
To better understand  the flow I had in mind:
![oauth-client-agent-flow](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/89905860/8251e43b-82bc-416b-9bef-74e786168548)

**Reply by DGuhr** [reply]
@jpkrohling Sorry to ping you, and feel free to ignore, but I'd really be interested in your comment on the topic above, as it is more about the extensions and their usage than the java agent itself, and you wrote the article I am referencing above. Regardless of an answer, have a nice day! :)

**Reply by jpkrohling** [reply]
> The article talks about 2 collectors if I read it now right

Yes, sorry that the terminology is a bit confusing in this world. The "agent" mentioned in the article is a Collector instance acting as a local collector (agent), sending data to a remote collector (collector). The Java Agent is an instrumentation agent, helping create the data that is then sent to collectors (local or remote). 

Other folks could probably help you on the Java side of things, but I don't think it can exchange access tokens into bearer tokens, nor refresh them periodically. If you have a bearer token that is valid for a long period, you can use the `OTEL_EXPORTER_OTLP_HEADERS` env var to get the OTLP exporter from the Java SDK to send the token to the collector: https://opentelemetry.io/docs/specs/otel/protocol/exporter/

**Reply by DGuhr** [reply]
@jpkrohling Thanks so much for clarifying :) I think about implementing an (java agent) extension now based on the diagram above, and if only to learn how to. It should be compatible with the oidc server (collector) extension, so if you have a hint what header needs to be sent for that extension to pick it up and validate it, that'd be nice. Guessing `"Authorization: Bearer <token-from-javaclient>"` here, but want to make sure :) Couldn't find the oidc/server go code yet (sry, the ecsosystem is just huge 😅)

**Reply by DGuhr** [reply]
Ok, I am at the point where I get clientid, secret and tokenendpoint using custom config params (my playground branch is [here](https://github.com/DGuhr/opentelemetry-java-instrumentation/tree/test_custom_extension), if interested). Easier than expected, though I have no clue if my way is idiomatic or not. 

Thanks again for both of you answering, I'll ask more concrete questions about implementing the custom extension as new questions, bc this one already is semi-offtopic for a while. 😇
